File: #0207.course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {

        vector<vector<int>> v1(numCourses, vector<int>());
        vector<int> v2(numCourses);
        for (auto v : prerequisites) {
            v1[v[1]].push_back(v[0]);
            ++v2[v[0]];
        }

        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (v2[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            int i = q.front();
            q.pop();
            for (auto v : v1[i]) {
                --v2[v];
                if (v2[v] == 0) {
                    q.push(v);
                }
            }
        }

        for (int i = 0; i < numCourses; ++i) {
            if (v2[i] != 0) {
                return false;
            }
        }
        return true;
    }
};

File: #0207.course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const graph = Array.from(Array(numCourses), () => []);
    const ranks = Array(numCourses).fill(0);
    prerequisites.forEach(([u, v]) => {
        graph[v].push(u);
        ranks[u]++;
    });

    const queue = [];
    ranks.forEach((r, i) => {
        if (!r) {
            queue.push(i);
        }
    });

    while (queue.length) {
        const f = queue.shift();
        numCourses--;
        graph[f].forEach((neigh) => {
            ranks[neigh]--;
            if (!ranks[neigh]) queue.push(neigh);
        });
    }
    return !numCourses;
};
